include/experimental/__p2630_bits/submdspan_mapping.hpp

```diff
@@ -0,0 +1,157 @@
+
+namespace std {
+namespace experimental {
+
+//******************************************
+// Return type of submdspan_mapping overloads
+//******************************************
+template <class Mapping> struct mapping_offset {
+  Mapping map;
+  size_t offset;
+};
+
+namespace detail {
+// constructs sub strides
+template <class SrcMapping, size_t... InvMapIdxs>
+constexpr auto construct_sub_strides(const SrcMapping &src_map,
+                                     index_sequence<InvMapIdxs...>) {
+  return array<typename SrcMapping::index_type, sizeof...(InvMapIdxs)>{
+      src_map.stride(InvMapIdxs)...};
+}
+} // namespace detail
+
+//**********************************
+// layout_left submdspan_mapping
+//*********************************
+namespace detail {
+
+// Figure out whether to preserve layout_left
+template <class IndexSequence, size_t SubRank, class... SliceSpecifiers>
+struct preserve_layout_left_mapping;
+
+template <class... SliceSpecifiers, size_t... Idx, size_t SubRank>
+struct preserve_layout_left_mapping<index_sequence<Idx...>, SubRank,
+                                    SliceSpecifiers...> {
+  constexpr static bool value =
+      (SubRank == 0) ||
+      ((Idx < SubRank - 1
+            ? is_same_v<SliceSpecifiers, full_extent_t>
+            : (Idx == SubRank - 1 ? is_same_v<SliceSpecifiers, full_extent_t> ||
+                                        is_convertible_v<SliceSpecifiers,
+                                                         tuple<size_t, size_t>>
+                                  : true)) &&
+       ...);
+};
```

Comment:
The expression for `value` is a bit hard to read, but it's also hard to simplify it, given that all the repeated `bool`s in it (like `is_same_v<SliceSpecifiers, full_extent_t>`) depend on a pack.  Rewriting them into `tuple`s or `array`s somehow makes it worse.

Would you consider changing the term
```c++
Idx == SubRank - 1 ? is_same_v<SliceSpecifiers, full_extent_t> ||
                                        is_convertible_v<SliceSpecifiers,
                                                         tuple<size_t, size_t>>
                                  : true)
```
into the following?
```c++
Idx != SubRank - 1 or
  is_same_v<SliceSpecifiers, full_extent_t> or
  is_convertible_v<SliceSpecifiers, tuple<size_t, size_t>>
```

Reply:
ah that is better :-)

Reply:
actually that expression is wrong ...
this thing will return true for integers and what not.

Reply:
I inserted a correct one which is more structured like this instead of nested ?: statements.